src/GraphBLAS-sharp.Backend/Algorithms/BFS.fs

```diff
@@ -11,68 +11,62 @@ module BFS =
         workGroupSize
         =
 
-        let transpose =
-            CSRMatrix.transpose clContext workGroupSize
-
-        let spMV = SpMV.run clContext add mul workGroupSize
+        let spMVTo =
+            SpMV.runTo clContext add mul workGroupSize
 
         let zeroCreate =
             ClArray.zeroCreate clContext workGroupSize
 
         let ofList = Vector.ofList clContext Dense
 
-        let maskComplemented =
-            DenseVector.DenseVector.elementWise clContext StandardOperations.complementedMaskOp workGroupSize
+        let maskComplementedTo =
+            DenseVector.DenseVector.elementWiseTo clContext StandardOperations.complementedMaskOp workGroupSize
 
-        let fillSubVector =
-            Vector.standardFillSubVector<int, int> clContext workGroupSize
+        let fillSubVectorTo =
+            DenseVector.DenseVector.standardFillSubVectorTo<int, int> clContext workGroupSize
 
-        let reduce =
-            Vector.reduce clContext workGroupSize addNumeric
+        let containsNonZero =
+            DenseVector.DenseVector.containsNonZero clContext workGroupSize
 
         fun (queue: MailboxProcessor<Msg>) (matrix: CSRMatrix<'a>) (source: int) ->
             let vertexCount = matrix.RowCount
 
-            let mutable levels: ClVector<int> =
-                zeroCreate queue vertexCount |> ClVectorDense
-
-            let mutable frontier = ofList vertexCount [ source, 1 ]
-            let transposedMatrix = transpose queue matrix
-
-            let mutable level = 0
-            let mutable stop = false
+            let levels = zeroCreate queue vertexCount
 
-            while not stop do
-                level <- level + 1
+            let frontier = ofList vertexCount [ source, 1 ]
 
-                let newLevels =
-                    fillSubVector queue levels frontier (clContext.CreateClCell level)
+            match frontier with
+            | ClVectorDense front ->
 
-                levels.Dispose queue
+                let mutable level = 0
+                let mutable stop = false
 
-                match frontier, newLevels with
-                | ClVectorDense f, ClVectorDense nl ->
-                    let newFrontierUnmasked = spMV queue transposedMatrix f
+                while not stop do
+                    level <- level + 1
 
-                    let newFrontier =
-                        maskComplemented queue newFrontierUnmasked nl
+                    //Assigning new level values
+                    fillSubVectorTo queue levels front (clContext.CreateClCell level) levels
+                    |> ignore
 
-                    newFrontierUnmasked.Dispose queue
-                    frontier.Dispose queue
+                    //Getting new frontier
+                    spMVTo queue matrix front front |> ignore
 
-                    frontier <- newFrontier |> ClVectorDense
-                    levels <- newLevels
+                    maskComplementedTo queue front levels front
```

Comment:
Should version which explicitly store result into provide buffer returns something?